### DISCUSSION TITLE
API for number of nodes and their weights in each layer for spaCy NER

### DISCUSSION THREAD

**Discussion by davesuketu215** [explosion/spaCy#7964] [Help: Coding & Implementations] [answered]
Is there any API or any function inside the spaCy standard code which returns the number of nodes in each layer of the neural network for a custom-trained spaCy NER model? This would be helpful in hyperparameter tuning/optimization.

Basically, I would like to use this information to decide on the optimum dropout rate. Also, another use is probably to get the weights of each node in each layer and then decide on the best value for grad_clip in the optimizer.

**Comment by svlandeg** [comment, marked as answer]
Hi! We have a `spacy debug model` command that provides you some insights into the internal network layers, e.g.:
```
 spacy debug model tagger_cpu.cfg tagger --paths.train="my_train_data" --paths.dev="my_dev_data" -P2 -DIM -PAR -GRAD
```
gives you
```
ℹ Layer 0: model ID 63: 'tok2vec-listener>>with_array(softmax)'
ℹ  - dim nO: 40
ℹ Layer 1: model ID 60: 'tok2vec-listener'
ℹ  - dim nO: 96
ℹ Layer 2: model ID 62: 'with_array(softmax)'
ℹ  - dim nO: 40
ℹ  - dim nI: 96
ℹ Layer 3: model ID 61: 'softmax'
ℹ  - dim nO: 40
ℹ  - dim nI: 96
ℹ  - param W: (40, 96) - sample: [ 0.00160265  0.00160261 -0.00160259
-0.00160266  0.00160268]
ℹ  - param b: (40,) - sample: [0.00160265 0.00160263 0.00160262 0.00160261
0.0016026 ]
ℹ  - grad W: (40, 96) - sample: [0. 0. 0. 0. 0.]
ℹ  - grad b: (40,) - sample: [0. 0. 0. 0. 0.]
```
If you want something custom, or you want to access this information from code, you can take some inspiration from the `_print_model` function defined by this CLI statement: https://github.com/explosion/spaCy/blob/master/spacy/cli/debug_model.py#L199
It basically traverses the nodes in the model with `model.walk()` and prints all information that's there:
```
    for i, node in enumerate(model.walk()):
        if not layers or i in layers:
            msg.info(f"Layer {i}: model ID {node.id}: '{node.name}'")

            if dimensions:
                for name in node.dim_names:
                    if node.has_dim(name):
                        msg.info(f" - dim {name}: {node.get_dim(name)}")
                    else:
                        msg.info(f" - dim {name}: {node.has_dim(name)}")

            if parameters:
                for name in node.param_names:
                    if node.has_param(name):
                        print_value = _print_matrix(node.get_param(name))
                        msg.info(f" - param {name}: {print_value}")
                    else:
                        msg.info(f" - param {name}: {node.has_param(name)}")
            if gradients:
                for name in node.param_names:
                    if node.has_grad(name):
                        print_value = _print_matrix(node.get_grad(name))
                        msg.info(f" - grad {name}: {print_value}")
                    else:
                        msg.info(f" - grad {name}: {node.has_grad(name)}")
            if attributes:
                attrs = node.attrs
                for name, value in attrs.items():
                    msg.info(f" - attr {name}: {value}")
```
Now, while typing this answer to your question, I noticed that there are some issues with the `debug model` command, especially when there's a `Transformer` in the pipeline. If you run into issues with it, feel free to ping me here, I'll probably be updating/fixing this command in the next few days.

Also let me know if this is the type of information you're looking for or not! I always find it fun to look at the details of the networks, so happy to help :-)
